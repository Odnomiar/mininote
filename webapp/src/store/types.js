const state = {
  SELECTED_NOTE_ID: 'selectedNoteId',
  NOTEBOOKS: 'notebooks',
  VERSION: 'version',
}
const getters = {
  LOADED: 'loaded',
  DIRTY: 'dirty',
  NOTES: 'notes',
  SELECTED_NOTE: 'selectedNote',
  NOTE_BY_ID: 'noteById',
  CURRENT_CONTENT: 'currentContent',
  CURRENT_TITLE: 'currentTitle',
}
const actions = {
  LOAD_NOTEBOOK: 'loadNotebook',
  LIST_NOTEBOOKS: 'listNotebooks',
  CREATE_NOTEBOOK: 'createNotebook',
  UPDATE_NOTEBOOK: 'updateNotebook',
  UPDATE_NOTE: 'updateNote',
  APPLY_CHANGES: 'applyChanges',
}
const mutations = {
  SELECT_NOTE: 'selectNote',
  SET_LOAD_NOTEBOOK: 'setLoadNotebook',
  SELECT_FIRST: 'selectFirst',
  ADD_CHANGE: 'addChange',
  REVERT_CHANGES: 'revertChanges',
  RESET: 'reset',
}

export { state, getters, actions, mutations }
