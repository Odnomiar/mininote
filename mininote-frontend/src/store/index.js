import Vue from 'vue'
import Vuex from 'vuex'

import api from '../api'

Vue.use(Vuex)

const emptyState = {
    notebook: {},
    notebooks: [],
    changes: {
        add: [],
        delete: [],
        update: {}
    },
    selectedNoteId: -1,
    loadNotebookId: '',
    version: require('../../package.json').version
}

/// GETTERS ///
function loaded(state) {
    return state.notebook.hasOwnProperty('notes')
}

function dirty(state) {
    return !!(state.changes.add.length || state.changes.delete.length || Object.keys(state.changes.update).length)
}

function selectedNote(state) {
    return resolveNotes(state).find(n => n.id === state.selectedNoteId)
}

function noteById(state) {
    return (id) => resolveNotes(state).find(n => n.id === id)
}

function resolveNotes(state) {
    let notes = []
    notes = notes.concat(state.notebook.notes || [])
    notes = notes.concat(state.changes.add)
    notes = notes.filter(n1 => !state.changes.delete.some(n2 => n1.id === n2.id))
    return notes
}

function currentContent(state) {
    let id = state.selectedNoteId
    if (id < 0) return ''
    if (state.changes.update.hasOwnProperty(id)) return state.changes.update[id].content
    return selectedNote(state).content
}

function currentTitle(state) {
    let id = state.selectedNoteId
    if (id < 0) return ''
    if (state.changes.update.hasOwnProperty(id)) return state.changes.update[id].title
    return selectedNote(state).title
}

/// MUTATIONS ///
function selectNote(state, id) {
    if (id < 0) return null
    let notes = resolveNotes(state)
    if (notes.some(n => n.id === id)) {
        Vue.set(state, 'selectedNoteId', id)
    }
}

function selectFirst(state) {
    let notes = resolveNotes(state)
    let id = notes.length ? notes[0].id : -1
    Vue.set(state, 'selectedNoteId', id)
}

/// STORE ///
const store = new Vuex.Store({
    state: JSON.parse(JSON.stringify(emptyState)),
    getters: {
        selectedNote,
        dirty,
        notes: resolveNotes,
        currentContent,
        currentTitle,
        loaded,
        noteById
    },
    mutations: {
        setNotebooks(state, notebookList) {
            Vue.set(state, 'notebooks', notebookList)
        },
        setNotebook(state, { id, password }) {
            Vue.set(state.notebook, 'id', id)
            Vue.set(state.notebook, 'password', password)
        },
        setNotes(state, notes) {
            Vue.set(state.notebook, 'notes', notes)
        },
        addNote(state, note) {
            Vue.set(state.notebook, 'notes', state.notebook.notes.concat([note]))
            if (state.selectedNoteId === note.oldId) selectNote(state, note.id)
        },
        removeNote(state, note) {
            Vue.set(state.notebook, 'notes', state.notebook.notes.filter(n => n.id !== note.id))
            if (state.selectedNoteId === note.id && state.notebook.notes.length) {
                Vue.set(state, 'selectedNoteId', state.notes[0].id)
            }
        },
        updateNote(state, note) {
            let notes = resolveNotes(state)
            let found = notes.find(n => n.id === note.id)
            for (let prop in found) {
                Vue.set(found, prop, note[prop])
            }
        },
        selectNote,
        selectFirst,
        setLoadNotebook(state, id) {
            Vue.set(state, 'loadNotebookId', id)
        },
        addChange(state, { type, payload }) {
            switch (type) {
                case 'add':
                case 'delete':
                    Vue.set(state.changes, type, state.changes[type].concat([payload]))
                    break
                case 'update':
                    let hasChange = state.changes.update.hasOwnProperty(payload.id)
                    let originalNote = (state.notebook.notes || []).find(n => n.id === payload.id)

                    if (originalNote && originalNote.content === payload.content && hasChange) {
                        Vue.delete(state.changes.update, payload.id)
                    } else {
                        Vue.set(state.changes.update, payload.id, payload)
                    }
                    break
            }
        },
        removeChange(state, { type, payload }) {
            switch (type) {
                case 'add':
                case 'delete':
                    let idx = state.changes[type].findIndex(n => n.id === payload.id)
                    Vue.delete(state.changes[type], idx)
                    break
                case 'update':
                    Vue.delete(state.changes.update, payload.id)
            }
        },
        revertChanges(state) {
            Vue.set(state.changes, 'add', [])
            Vue.set(state.changes, 'delete', [])

            Object.keys(state.changes.update).forEach(id => {
                let note = (state.notebook.notes || []).find(n => n.id === parseInt(id))
                let content = note.content

                Vue.set(note, 'content', '')
                Vue.set(note, 'content', content)
            })

            Vue.set(state.changes, 'update', {})
        },
        reset(state) {
            for (let prop in emptyState) {
                Vue.set(state, prop, JSON.parse(JSON.stringify(emptyState[prop])))
            }
        }
    },
    actions: {
        listNotebooks({ commit }) {
            return api.list()
                .then(data => commit('setNotebooks', data))
        },
        loadNotebook({ commit }, { id, password }) {
            return api.getNotes(id, password)
                .then(data => commit('setNotes', data))
                .then(() => commit('setNotebook', { id, password }))
        },
        createNotebook({ commit }, { id, password }) {
            return api.create(id, password)
                .then(data => commit('setNotes', data.notes))
                .then(() => commit('setNotebook', { id, password }))
        },
        updateNotebook({ state, commit }, notebook) {
            let updateNotebook = JSON.parse(JSON.stringify(notebook))
            if (notebook.id === state.notebook.id) delete updateNotebook.id
            return api.update(state.notebook.id, state.notebook.password, updateNotebook)
                .then(() => commit('setNotebook', notebook))
        },
        createNote({ state, commit }, note) {
            return api.addNote(state.notebook.id, state.notebook.password, note)
                .then(data => commit('addNote', { ...note, id: data.id, oldId: note.id }))
        },
        updateNote({ state, commit }, note) {
            return api.updateNote(state.notebook.id, state.notebook.password, note)
                .then(() => commit('updateNote', note))
        },
        deleteNote({ state, commit }, note) {
            return api.deleteNote(state.notebook.id, state.notebook.password, note)
                .then(() => commit('removeNote', note))
        },
        applyChanges({ state, commit, dispatch }) {
            let promises = []

            Object.values(state.changes.update).forEach(n => {
                let newNote = state.changes.add.find(n1 => n1.id === n.id)
                let deletedNote = state.changes.delete.find(n1 => n1.id === n.id)

                if (newNote) {
                    commit('updateNote', n)
                    commit('removeChange', { type: 'update', payload: n })
                }
                else if (deletedNote) commit('removeChange', { type: 'update', payload: n })
                else promises.push(dispatch('updateNote', n).then(() => commit('removeChange', { type: 'update', payload: n })))
            })

            state.changes.add.forEach(n => {
                if (state.changes.delete.some(n1 => n1.id === n.id)) {
                    commit('removeChange', { type: 'add', payload: n })
                    commit('removeChange', { type: 'delete', payload: n })
                }
                else promises.push(dispatch('createNote', n).then(() => commit('removeChange', { type: 'add', payload: n })))
            })
            state.changes.delete.forEach(n => {
                promises.push(dispatch('deleteNote', n).then(() => commit('removeChange', { type: 'delete', payload: n })))
            })

            return Promise.all(promises)
        }
    },
    strict: true
})

export default store