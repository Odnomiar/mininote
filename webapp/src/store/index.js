import Vuex from 'vuex'

import api from '../api'

import { version } from '../../package.json'

/* eslint-disable no-case-declarations */

const emptyState = {
  notebook: {},
  notebooks: [],
  changes: {
    add: [],
    delete: [],
    update: {},
  },
  selectedNoteId: -1,
  loadNotebookId: '',
  version,
}

/// GETTERS ///
function loaded(state) {
  return !!state.notebook.notes
}

function dirty(state) {
  return !!(
    state.changes.add.length ||
    state.changes.delete.length ||
    Object.keys(state.changes.update).length
  )
}

function selectedNote(state) {
  return resolveNotes(state).find((n) => n.id === state.selectedNoteId)
}

function noteById(state) {
  return (id) => resolveNotes(state).find((n) => n.id === id)
}

function resolveNotes(state) {
  let notes = []
  notes = notes.concat(state.notebook.notes || [])
  notes = notes.concat(state.changes.add)
  notes = notes.filter(
    (n1) => !state.changes.delete.some((n2) => n1.id === n2.id),
  )
  return notes
}

function currentContent(state) {
  let id = state.selectedNoteId
  if (id < 0) return ''
  if (id in state.changes.update) return state.changes.update[id].content
  const note = selectedNote(state)
  return note ? note.content : ''
}

function currentTitle(state) {
  let id = state.selectedNoteId
  if (id < 0) return ''
  if (id in state.changes.update) return state.changes.update[id].title
  const note = selectedNote(state)
  return note ? note.title : ''
}

/// MUTATIONS ///
function selectNote(state, id) {
  if (id < 0) state.selectedNoteId = -1
  let notes = resolveNotes(state)
  if (notes.some((n) => n.id === id)) {
    state.selectedNoteId = id
  }
}

function selectFirst(state) {
  let notes = resolveNotes(state)
  let id = notes.length ? notes[0].id : -1
  state.selectedNoteId = id
}

function updateUrl(state) {
  location.hash = state.notebook.id ? `#/notebook/${state.notebook.id}` : ''
}

/// STORE ///
const store = new Vuex.Store({
  state: JSON.parse(JSON.stringify(emptyState)),
  getters: {
    selectedNote,
    dirty,
    notes: resolveNotes,
    currentContent,
    currentTitle,
    loaded,
    noteById,
  },
  mutations: {
    setNotebooks(state, notebookList) {
      state.notebooks = notebookList
    },
    setNotebook(state, { id, password }) {
      state.notebook.id = id
      state.notebook.password = password
      updateUrl(state)
    },
    setNotes(state, notes) {
      state.notebook.notes = notes
    },
    addNote(state, note) {
      state.notebook.notes = state.notebook.notes.concat([note])
      if (state.selectedNoteId === note.oldId) selectNote(state, note.id)
    },
    removeNote(state, note) {
      state.notebook.notes = state.notebook.notes.filter(
        (n) => n.id !== note.id,
      )
      if (state.selectedNoteId === note.id && state.notebook.notes.length) {
        state.selectedNoteId = state.notes[0].id
      }
    },
    updateNote(state, note) {
      let notes = resolveNotes(state)
      let found = notes.find((n) => n.id === note.id)
      for (let prop in found) {
        found[prop] = note[prop]
      }
    },
    selectNote,
    selectFirst,
    setLoadNotebook(state, id) {
      state.loadNotebookId = id
    },
    addChange(state, { type, payload }) {
      switch (type) {
        case 'add':
        case 'delete':
          state.changes[type] = state.changes[type].concat([payload])
          break
        case 'update':
          const hasChange = payload.id in state.changes.update
          const originalNote = (state.notebook.notes || []).find(
            (n) => n.id === payload.id,
          )

          if (
            originalNote &&
            originalNote.content === payload.content &&
            hasChange
          ) {
            delete state.changes.update[payload.id]
          } else {
            state.changes.update[payload.id] = payload
          }
          break
      }
    },
    removeChange(state, { type, payload }) {
      switch (type) {
        case 'add':
        case 'delete':
          state.changes[type] = state.changes[type].filter(
            (n) => n.id !== payload.id,
          )
          break
        case 'update':
          delete state.changes.update[payload.id]
      }
    },
    revertChanges(state) {
      state.changes.add = []
      state.changes.delete = []

      Object.keys(state.changes.update).forEach((id) => {
        let note = (state.notebook.notes || []).find(
          (n) => n.id === parseInt(id),
        )

        if (!note) return

        let content = note.content
        note.content = ''
        note.content = content
      })

      state.changes.update = {}

      if (!resolveNotes(state).length) {
        selectFirst(state) // select none, i.e. -1
      }
    },
    reset(state) {
      for (let prop in emptyState) {
        state[prop] = JSON.parse(JSON.stringify(emptyState[prop]))
      }
      updateUrl(state)
    },
  },
  actions: {
    listNotebooks({ commit }) {
      return api.list().then((data) => commit('setNotebooks', data))
    },
    loadNotebook({ commit }, { id, password }) {
      return api
        .getNotes(id, password)
        .then((data) => commit('setNotes', data))
        .then(() => commit('setNotebook', { id, password }))
    },
    createNotebook({ commit }, { id, password }) {
      return api
        .create(id, password)
        .then((data) => commit('setNotes', data.notes))
        .then(() => commit('setNotebook', { id, password }))
    },
    updateNotebook({ state, commit }, notebook) {
      let updateNotebook = JSON.parse(JSON.stringify(notebook))
      if (notebook.id === state.notebook.id) delete updateNotebook.id
      return api
        .update(state.notebook.id, state.notebook.password, updateNotebook)
        .then(() => commit('setNotebook', notebook))
    },
    createNote({ state, commit }, note) {
      return api
        .addNote(state.notebook.id, state.notebook.password, note)
        .then((data) =>
          commit('addNote', { ...note, id: data.id, oldId: note.id }),
        )
    },
    updateNote({ state, commit }, note) {
      return api
        .updateNote(state.notebook.id, state.notebook.password, note)
        .then(() => commit('updateNote', note))
    },
    deleteNote({ state, commit }, note) {
      return api
        .deleteNote(state.notebook.id, state.notebook.password, note)
        .then(() => commit('removeNote', note))
    },
    applyChanges({ state, commit, dispatch }) {
      let promises = []

      Object.values(state.changes.update).forEach((n) => {
        let newNote = state.changes.add.find((n1) => n1.id === n.id)
        let deletedNote = state.changes.delete.find((n1) => n1.id === n.id)

        if (newNote) {
          commit('updateNote', n)
          commit('removeChange', { type: 'update', payload: n })
        } else if (deletedNote)
          commit('removeChange', { type: 'update', payload: n })
        else
          promises.push(
            dispatch('updateNote', n).then(() =>
              commit('removeChange', { type: 'update', payload: n }),
            ),
          )
      })

      state.changes.add.forEach((n) => {
        if (state.changes.delete.some((n1) => n1.id === n.id)) {
          commit('removeChange', { type: 'add', payload: n })
          commit('removeChange', { type: 'delete', payload: n })
        } else
          promises.push(
            dispatch('createNote', n).then(() =>
              commit('removeChange', { type: 'add', payload: n }),
            ),
          )
      })
      state.changes.delete.forEach((n) => {
        promises.push(
          dispatch('deleteNote', n).then(() =>
            commit('removeChange', { type: 'delete', payload: n }),
          ),
        )
      })

      return Promise.all(promises)
    },
  },
  strict: true,
})

export function useStore() {
  return store
}
