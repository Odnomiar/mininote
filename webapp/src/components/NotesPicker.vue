<template>
  <div class="picker-container">
    <core-modal
      v-if="showRenameModal"
      @ok="_updateNote"
      @discard="showRenameModal = false"
    >
      <template #header>Rename Note</template>
      <template #main>
        <div class="flex flex-col">
          <div class="flex">
            <div class="w-1/4">
              <label for="newNameInput">Name:</label>
            </div>
            <div class="w-3/4">
              <input
                id="newNameInput"
                v-model="inputs.newName"
                type="text"
                placeholder="Enter new name for this note."
                class="w-full bg-gray-100 focus:shadow-outline"
                style="border: 1px solid #cbd5e0"
              />
            </div>
          </div>
        </div>
      </template>
      <template #ok-btn>Save</template>
    </core-modal>

    <div class="flex mb-8">
      <span
        class="flex items-center px-4 bg-gray-300 border-r border-gray-400 rounded-l-md"
        >&#x1F50D;</span
      >
      <input
        v-model="query"
        type="text"
        class="placeholder-gray-700 rounded-l-none"
        placeholder="Search ..."
      />
    </div>

    <ul>
      <li
        v-for="n in filteredNotes"
        :key="n.id"
        class="flex items-center justify-between px-4 py-2 bg-gray-100 cursor-pointer first:rounded-t-md last:rounded-b-md"
        :class="{ 'note-active': selectedNoteId == n.id }"
        @click="selectNote(n.id)"
      >
        {{ n.title }}
        <div class="flex flex-row-reverse">
          <span class="ml-1 delete-btn" @click.stop="_deleteNote(n)">üóë</span>
          <span class="mr-1 rename-btn" @click="showRenameModal = true"
            >‚úèÔ∏è</span
          >
        </div>
      </li>
    </ul>
    <div class="flex mt-8">
      <input
        v-model="newNoteInput"
        type="text"
        class="placeholder-gray-700 rounded-r-none"
        placeholder="Add note ..."
        @keyup.enter="_addNote"
      />
      <button
        type="button"
        class="px-4 rounded-l-none btn-primary"
        :disabled="!newNoteInput"
        @click="_addNote"
      >
        &#x2b;
      </button>
    </div>
  </div>
</template>

<script>
  import { mapActions, mapGetters, mapMutations, mapState } from 'vuex'
  import { actions, getters, mutations } from '@/store/types'

  import CoreModal from '@/components/core/CoreModal'

  export default {
    components: {
      CoreModal,
    },
    emits: ['alert', 'notebook-selected'],
    data() {
      return {
        query: '',
        newNoteInput: '',
        inputs: {
          newName: '',
        },
        showRenameModal: false,
      }
    },
    computed: {
      ...mapState({
        notebook: (state) => state.notebook,
        selectedNoteId: (state) => state.selectedNoteId,
      }),
      ...mapGetters([
        getters.NOTES,
        getters.CURRENT_TITLE,
        getters.SELECTED_NOTE,
      ]),
      filteredNotes: function () {
        let filtered =
          this.query === ''
            ? [...this.notes]
            : this.notes.filter((n) => {
                let pattern = new RegExp(this.query, 'gi')
                return pattern.test(n.title)
              })
        return filtered.sort((el1, el2) =>
          el1.title < el2.title ? -1 : el1.title > el2.title ? 1 : 0,
        )
      },
    },
    watch: {
      selectedNoteId: function () {
        this.inputs.newName = this.currentTitle
      },
    },
    mounted() {
      this.selectFirst()
    },
    methods: {
      ...mapMutations([
        mutations.SELECT_NOTE,
        mutations.SELECT_FIRST,
        mutations.ADD_CHANGE,
      ]),
      ...mapActions([actions.UPDATE_NOTE]),
      handleError: function (err) {
        this.$emit('alert', err.message)
        this.reset()
      },
      reset: function () {
        this.newNoteInput = ''
      },
      _addNote: function () {
        if (!this.newNoteInput) return

        let note = {
          id: this.notes.reduce((acc, n) => Math.max(acc, n.id), 0) + 1024 + 1,
          title: this.newNoteInput,
          content: '',
        }

        this.addChange({ type: 'add', payload: note })
        this.selectNote(note.id)
        this.newNoteInput = ''
      },
      _updateNote: function () {
        this.updateNote({
          ...this.selectedNote,
          title: this.inputs.newName,
        })
          .then(() => {
            this.$emit('alert', 'Note saved.', 'success')
            this.showRenameModal = false
          })
          .catch(this.handleError)
      },
      _deleteNote: function (note) {
        this.addChange({ type: 'delete', payload: note })
        if (this.selectedNoteId === note.id) {
          this.selectFirst()
        }
      },
    },
  }
</script>

<style scoped>
  .note-active {
    @apply bg-green-600 text-gray-100;
  }
</style>
