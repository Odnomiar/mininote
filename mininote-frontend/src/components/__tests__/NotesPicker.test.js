import { createLocalVue, shallowMount } from '@vue/test-utils'
import BootstrapVue from 'bootstrap-vue'
import NotesPicker from '@/components/NotesPicker'
import store from '../../store'

const localVue = createLocalVue()

localVue.use(BootstrapVue)

describe('NotesPicker.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(NotesPicker, {
      localVue,
      propsData: {
        notes: ['note one', 'note two']
      },
      store
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
