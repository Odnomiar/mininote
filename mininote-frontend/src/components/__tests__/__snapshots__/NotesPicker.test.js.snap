// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesPicker.vue matches snapshot 1`] = `
<div class="picker-container" notes="note one,note two">
  <div class="search-container">
    <div class="input-group">
      <div id="sizing-addon1" class="input-group-prepend"><span class="input-group-text">🔍</span></div> <input type="text" placeholder="Search ..." class="form-control">
    </div>
  </div>
  <ul class="list-group"></ul>
  <div class="input-group new-note-input"><input type="text" placeholder="Add note ..." class="form-control">
    <div class="input-group-btn"><button type="button" disabled="disabled" class="btn btn-primary">+</button></div>
  </div>
</div>
`;
